File: 2023/day-01/day-01.c
#include "stdio.h"

#define TRUE 1
#define FALSE 0

// https://adventofcode.com/2023/day/1

char *validDigits[] = {
    "one", "two", "three", "four", "five", "six", "seven", "eight", "nine"
};

int getDigitValue(char *str, int start, int maxLength) {
    char c = str[start];
    if (c >= '0' && c <= '9') {
        return c - '0';
    }
    if (maxLength < 3) {
        return -1;
    }
    for (int i = 0; i < 9; i++) {
        char *valid_digit = validDigits[i];
        int valid_digit_length = 0;
        while (valid_digit[valid_digit_length] != '\0') {
            valid_digit_length++;
        }
        for (int j = 0; j < valid_digit_length; j++) {
            if (start + j >= maxLength) {
                break;
            }
            if (str[start + j] != valid_digit[j]) {
                break;
            }
            if (j == valid_digit_length - 1) {
                return i + 1;
            }
        }
    }
    return -1;
}

int day01(char *filename) {
    FILE *f = fopen(filename, "r");
    if (f == NULL) {
        printf("Could not open file %s\n", filename);
        return -1;
    }
    int sum = 0;
    char line[100];
    while (fgets(line, 100, f) != NULL) {
        int digit1Found = FALSE;
        int digit2Found = FALSE;
        int digit1 = 0;
        int digit2 = 0;
        int lineLength = 0;
        while (line[lineLength] != '\0') {
            lineLength++;
        }
        for (int i = 0; (!digit1Found || !digit2Found) && i < lineLength; i++) {
            int digit_value = getDigitValue(line, i, lineLength);
            if (digit_value >= 0) {
                if (!digit1Found) {
                    digit1Found = TRUE;
                    digit1 = digit_value;
                }
            }
            digit_value = getDigitValue(line, lineLength - i - 1, lineLength);
            if (digit_value >= 0) {
                if (!digit2Found) {
                    digit2Found = TRUE;
                    digit2 = digit_value;
                }
            }
        }
        sum = sum + digit1 * 10 + digit2;
    }
    fclose(f);

    return sum;
}

int main(int argc, char *argv[]) {
    char *filename = "input.txt";
    if (argc > 1) {
        filename = argv[1];
    }
    int result = day01(filename);
    if (result < 0) {
        return -1;
    }
    printf("The sum of all of the calibration values is %d\n", result);
    return 0;
}

File: 2023/day-01/day-01.rs
use std::env;
use std::fs;

const VALID_DIGITS: [&str; 9] = [
    "one", "two", "three", "four", "five", "six", "seven", "eight", "nine",
];

fn get_valid_digit(search: &str) -> i32 {
    let first_char = search.chars().nth(0).unwrap();
    if first_char >= '0' && first_char <= '9' {
        return first_char as i32 - '0' as i32;
    }
    for (i, valid_digit) in VALID_DIGITS.iter().enumerate() {
        if search.starts_with(valid_digit) {
            return (i + 1) as i32;
        }
    }

    return -1;
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let file_path: &str;

    if args.len() > 1 {
        file_path = &args[1];
    } else {
        file_path = "input.txt";
    }

    let file = fs::read_to_string(file_path);

    if let Ok(contents) = file {
        let mut sum = 0;
        for line in contents.lines() {
            let mut digit1_found = false;
            let mut digit2_found = false;
            let mut digit1 = 0;
            let mut digit2 = 0;
            for (i, _) in line.chars().enumerate() {
                if !digit1_found {
                    let search_str = line.get(i..line.len()).unwrap();
                    let digit_value = get_valid_digit(search_str);
                    digit1_found = digit_value >= 0;
                    digit1 = digit_value;
                }
                if !digit2_found {
                    let search_str = line.get(line.len() - i - 1..line.len()).unwrap();
                    let digit_value = get_valid_digit(search_str);
                    digit2_found = digit_value >= 0;
                    digit2 = digit_value;
                }
            }
            if digit1_found && digit2_found {
                sum += digit1 * 10 + digit2;
            }
        }
        println!("The sum of all of the calibration values is {}", sum);
    }
}
